x/ante/check_refund.go

```diff
@@ -35,62 +36,69 @@ func NewCheckRefundFeeDecorator(registry cdctypes.InterfaceRegistry, ak antetype
 func (d CheckRefundFeeDecorator) AnteHandle(ctx sdk.Context, tx sdk.Tx, simulate bool, next sdk.AnteHandler) (sdk.Context, error) {
 	msgs := tx.GetMsgs()
 
-	// reject unregistered refundable msg
-	for _, msg := range msgs {
-		switch msg := msg.(type) {
-		case *rewardtypes.RefundMsgRequest:
-			if !msgRegistered(d.registry, msg.InnerMessage.GetTypeUrl()) {
-				return ctx, sdkerrors.Wrap(sdkerrors.ErrInvalidRequest, fmt.Sprintf("%T is not refundable", msg))
-			}
-		}
+	if !anyRefundable(msgs) {
+		return next(ctx, tx, simulate)
 	}
 
-	if d.qualifyForRefund(ctx, msgs) {
-		feeTx, ok := tx.(sdk.FeeTx)
-		if !ok {
-			return ctx, sdkerrors.Wrap(sdkerrors.ErrTxDecode, "Tx must be a FeeTx")
-		}
+	if err := d.qualifyForRefund(ctx, msgs); err != nil {
+		return ctx, err
+	}
 
-		fee := feeTx.GetFee()
-		if len(fee) > 0 {
-			req := msgs[0].(*rewardtypes.RefundMsgRequest)
-			err := d.reward.SetPendingRefund(ctx, *req, fee[0])
-			if err != nil {
-				return ctx, sdkerrors.Wrapf(sdkerrors.ErrInvalidRequest, err.Error())
-			}
-		}
+	feeTx, ok := tx.(sdk.FeeTx)
+	if !ok {
+		return ctx, sdkerrors.Wrap(sdkerrors.ErrTxDecode, "tx must be a FeeTx")
+	}
 
+	fee := feeTx.GetFee()
+	if len(fee) > 0 {
+		req := msgs[0].(*rewardtypes.RefundMsgRequest)
+		err := d.reward.SetPendingRefund(ctx, *req, fee[0])
```

Comment:
Unrelated question, when can fee have more than 1 element?

Reply:
you can pay fees in different denominations